main/guides/zoe/contract-details.md

```diff
@@ -0,0 +1,121 @@
+
+The Exo objects (or just "Exos") defined by these functions are both persistent and virtualizable. SwingSet
+knows their complete state, so it can page them out when space is tight, and page them back in when they are
+referenced again.
+
+Fields of the object returned by the `init` function become fields of the persistent state. (These cannot
+currently be changed on upgrade, though we're considering relaxing that restriction.) Within methods they can
+each be accessed as fields of `this.state`. Our convention is to extract the fields that will be used in a
+method on its first line, like `const { a, b } = this.state;` Once that has been done, those variable can be
+read or written as normal javascript variables, and the values will persist. (For more details, see [the note
+here](https://docs.agoric.com/guides/zoe/contract-upgrade.html#kinds)).
+
+### Methods: defining behavior
+
+The methods argument of `exoClass()` is a bag of methods, written in [concise method
+syntax](https://github.com/Agoric/agoric-sdk/wiki/Arrow-Function-Style#far-classes-do-not-use-arrow-function-style).
+`exoClass()` defines a single facet.
+
+The methodKit parameter to `exoClassKit` is a record of labelled bags of methods. Each label defines a facet
+of the object. All facets of each object share access to the same state, but each facet is a separate
+capability. Within the methods, other facets can be reached by name within `this.facets`. The maker you get
+from `exoClassKit()` builds a new object each time it is called, and return all the facets. The caller can
+decide which of the facets to pass to each recipient.
+
+### Guards: defensive methods
+
+These objects and facets are designed to be durable and shareable across address space boundaries. Since code
+in different vats might not be mutually trusting, code needs to be defensive about parameters received from
+remote callers. Interface Guards let us express in code what parameters each method expects and can handle
+safely. If a caller provides a parameter that doesn't match the template specified in the guard, SwingSet
+returns an exception to the caller without notifying the receiver. If the return value doesn't match, the
+function returns an exception to the caller.
+
+`exoClass()` takes a guard for a single interface, defined by
+
+```
+M.interface('name', {
+  methodA: M.call(paramsAShape).returns(resultAShape),
+  methodB: M.callWhen(M.await(paramsBShape)).returns(resultBShape),
+}
+```
+
+`M.call()` verifies that all parameters match the guard before passing them through to the
+method. `M.callWhen(M.await(paramsBGuard))` awaits the resolution of the parameter, and then verifies that the
+result matches before invoking the method. When a guard is written this latter way, the method doesn't have to
+be `async`. In both cases, the method doesn't have to check for compliance with the guard.
+
+[Shapes can specify](https://endojs.github.io/endo/interfaces/_endo_patterns.PatternMatchers.html) simple
```

Comment:
I made a patterns cheat-sheet at https://docs.agoric.com/reference/zoe-api/zoe-contract-facet.html#proposal-shapes, but that context is specialized to proposal shapes. hm.

Perhaps using glossary entries for Pattern and Shape would help. Perhaps I should do that in a follow-up PR.